_posts/2023-06-13-http4s-authentication-part2.md

```diff
@@ -0,0 +1,894 @@
+end BarCodeService
+```
+
+### 3.1.2 Creating the OtpService
+
+Let's create a scala file and save it as `OtpService.scala`. This service will contain the implementation for `Hotp` and `Totp` that we will use for this application.
+
+```scala
+package com.rockthejvm
+
+import com.bastiaanjansen.otp.*
+
+class OtpService(generator: String, id: Int) extends MyDatabase with BarCodeService:
+    private val secret: Array[Byte] = SecretGenerator.generate()
+end OtpService
+```
+
+We create it as a class in order to pass some arguments to the constructor during instantiation. As mentioned before, `generator` is a string of either `Hotp` or `Totp` while `id` points to a user in the database. This class extends `MyDatabase` and `BarCodeService` traits since we will make use of some of their functions.
+
+We first create our `secret` as a private val using `otp-java`, since we will only be using it within this class
+
+```scala
+import java.time.Duration
+
+class OtpService(generator: String, id: Int) extends MyDatabase with BarCodeService:
+    ...
+    enum Generator:
+        case Hotp, Totp
+
+        def getHotp: HOTPGenerator = new HOTPGenerator.Builder(secret)
+            .withPasswordLength(6)
+            .withAlgorithm(HMACAlgorithm.SHA1)
+            .build()
+
+        def getTotp: TOTPGenerator = new TOTPGenerator.Builder(secret)
+            .withHOTPGenerator(builder => {
+                    builder.withPasswordLength(6)
+                    builder.withAlgorithm(HMACAlgorithm.SHA1)
+            })
+            .withPeriod(Duration.ofSeconds(30))
+            .build()
+    end Generator
+end OtpService
+```
+
+The `Generator` enum is what we will use to choose between `Hotp` and `Totp` implementations depending on our needs, it also contains two methods, `getHotp`, and `getTotp` used to generate `HOTPGenerator` and `TOTPGenerator` instances respectfully. Note that Google Authenticator displays a token length of 6 and uses the `SHA1` algorithm. We supply these same values as we generate the `HOTPGenerator` and `TOTPGenerator` instances.
+
+```scala
+class OtpService(generator: String, id: Int) extends MyDatabase with BarCodeService:
+    ...
+    def getToken(generator: String = generator, id: Int = id): String = Generator.valueOf(generator) match
+        case Generator.Hotp => Generator.Hotp.getHotp.generate(getCounterValue(id))
+        case Generator.Totp => Generator.Totp.getTotp.now()
+end OtpService
+```
+
+The `getToken()` function takes two optional arguments, a string, `generator` and an Int, `id` both of which are passed through the `OtpService` constructor. `generator` is passed to the `Generator.valeOf` enum method which yields either an `Hotp` or `Totp` `Generator` case. We pattern match on this to give us a token value.
+
+```scala
+class OtpService(generator: String, id: Int) extends MyDatabase with BarCodeService:
+    ...
+    private def verifyUtility (value: Try[Boolean], id: Int = id): String =
```

Comment:
If we remove the need for Try, this function becomes redundant.
The pattern match for `true` and `false` can also be replaced by an if/else.

Reply:
Fixed